File: debug_daily_summary_comprehensive.js
const admin = require('firebase-admin');

// Initialize Firebase Admin
if (!admin.apps.length) {
  admin.initializeApp({
    projectId: 'plan-with-hands',
  });
}

const db = admin.firestore();
// Use the correct Firestore database (not the default)
db.settings({ databaseId: 'planwithhands' });

async function comprehensiveDailySummaryDebug() {
  console.log('🔍 === COMPREHENSIVE DAILY SUMMARY DEBUG ===\n');
  
  try {
    // 1. Check organizations and basic data
    console.log('📋 Step 1: Organization Data');
    const orgsSnapshot = await db.collection('organizations').get();
    console.log(`   Found ${orgsSnapshot.size} organizations`);
    
    if (orgsSnapshot.empty) {
      console.log('❌ No organizations found - this explains why no summaries are sent!');
      return;
    }
    
    for (const orgDoc of orgsSnapshot.docs) {
      const orgId = orgDoc.id;
      const orgData = orgDoc.data();
      console.log(`\n   📄 Organization: ${orgData.name || 'Unnamed'} (${orgId})`);
      
      // 2. Check admin users for this org
      console.log(`   👥 Admin Users Check:`);
      const adminUsers = await db.collection('users')
        .where('organizationId', '==', orgId)
        .where('userRole', 'in', [1, 2])
        .where('isActive', '==', true)
        .get();
      
      console.log(`      Found ${adminUsers.size} admin/manager users`);
      adminUsers.forEach(doc => {
        const data = doc.data();
        console.log(`      - ${data.firstName || ''} ${data.lastName || ''} (Role: ${data.userRole})`);
      });
      
      if (adminUsers.size === 0) {
        console.log('      ⚠️  NO ADMIN USERS - This would prevent summaries from being sent!');
        continue;
      }
      
      // 3. Check locations and timezones
      console.log(`   📍 Locations & Timezones:`);
      const locationsSnapshot = await db.collection('organizations')
        .doc(orgId)
        .collection('locations')
        .get();
      
      console.log(`      Found ${locationsSnapshot.size} locations`);
      const timezones = new Set();
      
      locationsSnapshot.forEach(doc => {
        const data = doc.data();
        const timezone = data.timezone;
        const locationName = data.locationName || 'Unknown';
        console.log(`      - ${locationName}: ${timezone || 'NO TIMEZONE SET'}`);
        if (timezone) timezones.add(timezone);
      });
      
      if (timezones.size === 0) {
        console.log('      ⚠️  NO TIMEZONES SET - This might affect scheduling logic!');
      } else {
        console.log(`      Unique timezones: ${Array.from(timezones).join(', ')}`);
      }
      
      // 4. Check recent daily summary logs
      console.log(`   📅 Recent Daily Summary Logs:`);
      const recentDates = ['2025-09-19', '2025-09-18', '2025-09-17'];
      
      for (const dateStr of recentDates) {
        const logDoc = await db.collection('organizations')
          .doc(orgId)
          .collection('daily_summary_logs')
          .doc(dateStr)
          .get();
        
        if (logDoc.exists) {
          const logData = logDoc.data();
          console.log(`      ✅ ${dateStr}: Sent at ${logData.sentAt?.toDate()}`);
        } else {
          console.log(`      ❌ ${dateStr}: No summary sent`);
        }
      }
      
      // 5. Check if there's task data for today
      console.log(`   📋 Task Data Check (2025-09-19):`);
      try {
        // Check for daily checklists across all locations
        let totalChecklists = 0;
        let totalTasks = 0;
        
        for (const locationDoc of locationsSnapshot.docs) {
          const checklistsSnapshot = await db.collection('organizations')
            .doc(orgId)
            .collection('locations')
            .doc(locationDoc.id)
            .collection('daily_checklists')
            .where('date', '==', '2025-09-19')
            .get();
          
          totalChecklists += checklistsSnapshot.size;
          
          for (const checklistDoc of checklistsSnapshot.docs) {
            const checklistData = checklistDoc.data();
            const tasks = checklistData.tasks || [];
            totalTasks += tasks.length;
            
            // Also check subcollection tasks
            const subTasks = await checklistDoc.ref.collection('tasks').get();
            totalTasks += subTasks.size;
          }
        }
        
        console.log(`      Found ${totalChecklists} checklists with ${totalTasks} total tasks`);
        
        if (totalTasks === 0) {
          console.log('      ⚠️  NO TASK DATA - This would prevent summaries (empty content check)');
        }
        
      } catch (error) {
        console.log(`      ❌ Error checking task data: ${error.message}`);
      }
      
      // 6. Check recent notifications for admin users
      if (adminUsers.size > 0) {
        console.log(`   📬 Recent Notifications:`);
        const firstAdminId = adminUsers.docs[0].id;
        try {
          const notifications = await db.collection('userNotifications')
            .doc(firstAdminId)
            .collection('notifications')
            .where('type', '==', 'daily_summary')
            .orderBy('createdAt', 'desc')
            .limit(3)
            .get();
          
          console.log(`      Found ${notifications.size} recent daily summary notifications`);
          notifications.forEach(doc => {
            const data = doc.data();
            console.log(`      - ${data.createdAt?.toDate()}: ${data.title}`);
          });
          
        } catch (error) {
          console.log(`      ❌ Error checking notifications: ${error.message}`);
        }
      }
    }
    
    // 7. Analyze the timezone logic issue
    console.log('\n⏰ Step 2: Timezone Logic Analysis');
    console.log('   Current UTC time:', new Date().toISOString());
  console.log('   Schedule model: every 5 minutes (*/5), UTC');
  console.log('   In-window logic: local hour == target hour AND');
  console.log('                    local minute in [targetMinute, targetMinute+5)');
  console.log('   Notes: Align org setting to a 5-minute boundary to trigger.');
    
    // 8. Function deployment verification
    console.log('\n🔧 Step 3: Function Deployment Status');
    console.log('   From previous check, these functions are deployed:');
  console.log('   ✅ scheduledDailySummary (scheduled, polls every 5 minutes)');
    console.log('   ✅ triggerDailySummary (callable)');
    console.log('   ✅ scheduledDailyGenerator (scheduled)');
    
    // 9. Recommendations
    console.log('\n🎯 Step 4: Recommendations');
    console.log('   Based on this analysis, likely issues:');
    console.log('   ');
    console.log('   1. 🚨 TIMEZONE LOGIC TOO RESTRICTIVE');
    console.log('      - Current logic prevents execution for most US timezones');
    console.log('      - Recommend: Remove timezone restriction or use better logic');
    console.log('   ');
    console.log('   2. ⚠️  Missing timezones in location data');
    console.log('      - Some locations have no timezone set');
    console.log('      - This defaults to UTC, which may not be intended');
    console.log('   ');
    console.log('   3. 📊 Check if task data exists');
    console.log('      - Function skips if no meaningful content found');
    console.log('      - Verify there are completed tasks to summarize');
    console.log('   ');
    console.log('   4. 👥 Verify admin users exist');
    console.log('      - Function skips if no admin users to send to');
    console.log('      - Ensure userRole field is set correctly (1=manager, 2=admin)');
    
  } catch (error) {
    console.error('❌ Fatal error:', error);
  }
}

comprehensiveDailySummaryDebug().then(() => {
  console.log('\n✅ Comprehensive debug completed');
  process.exit(0);
}).catch(error => {
  console.error('💥 Debug failed:', error);
  process.exit(1);
});

File: debug_daily_summary_comprehensive.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:hands_app/services/daily_summary_service.dart';
import 'package:hands_app/services/daily_background_service.dart';
import 'package:hands_app/utils/firestore_enforcer.dart';

/// Comprehensive debug script to investigate daily summary issues
Future<void> main() async {
  print('🔍 Starting comprehensive daily summary debug...');

  final firestore = FirestoreEnforcer.instance;
  final summaryService = DailySummaryService();

  try {
    // 1. Check if there are any organizations
    print('\n📊 Step 1: Checking organizations...');
    final orgsQuery = await firestore.collection('organizations').limit(5).get();
    print('Found ${orgsQuery.docs.length} organizations');

    for (final orgDoc in orgsQuery.docs) {
      final orgId = orgDoc.id;
      final orgData = orgDoc.data();
      final orgName = orgData['name'] ?? orgData['organizationName'] ?? 'Unknown Org';

      print('\n🏢 Organization: $orgName (ID: $orgId)');
      print('   Timezone: ${orgData['timezone'] ?? 'NOT SET'}');

      // Check admin users
      final adminQuery =
          await firestore
              .collection('users')
              .where('organizationId', isEqualTo: orgId)
              .where('userRole', whereIn: [1, 2])
              .where('isActive', isEqualTo: true)
              .get();

      print('   Admin users: ${adminQuery.docs.length}');
      for (final adminDoc in adminQuery.docs) {
        final adminData = adminDoc.data();
        print('     - ${adminData['firstName']} ${adminData['lastName']} (Role: ${adminData['userRole']})');
      }

      // Check locations
      final locQuery = await firestore.collection('organizations').doc(orgId).collection('locations').limit(3).get();

      print('   Locations: ${locQuery.docs.length}');
      for (final locDoc in locQuery.docs) {
        final locData = locDoc.data();
        final locName = locData['locationName'] ?? 'Unknown Location';
        final timezone = locData['timezone'];
        print('     - $locName (Timezone: ${timezone ?? 'NOT SET'})');
      }

      // Check recent daily summary logs
      final now = DateTime.now();
      final last7Days = List.generate(7, (i) => now.subtract(Duration(days: i)));

      print('   Daily summary logs (last 7 days):');
      for (final date in last7Days) {
        final dateStr =
            '${date.year.toString().padLeft(4, '0')}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
        final logDoc =
            await firestore.collection('organizations').doc(orgId).collection('daily_summary_logs').doc(dateStr).get();

        if (logDoc.exists) {
          final logData = logDoc.data()!;
          final sentAt = logData['sentAt'] as Timestamp?;
          print('     ✅ $dateStr - Sent at: ${sentAt?.toDate()}');
        } else {
          print('     ❌ $dateStr - No summary sent');
        }
      }

      // Check recent daily checklists
      final yesterdayStr =
          '${(now.subtract(const Duration(days: 1))).year.toString().padLeft(4, '0')}-${(now.subtract(const Duration(days: 1))).month.toString().padLeft(2, '0')}-${(now.subtract(const Duration(days: 1))).day.toString().padLeft(2, '0')}';

      int totalChecklists = 0;
      for (final locDoc in locQuery.docs) {
        final checklistQuery =
            await firestore
                .collection('organizations')
                .doc(orgId)
                .collection('locations')
                .doc(locDoc.id)
                .collection('daily_checklists')
                .where('date', isEqualTo: yesterdayStr)
                .get();
        totalChecklists += checklistQuery.docs.length;
      }
      print('   Daily checklists for yesterday ($yesterdayStr): $totalChecklists');

      // Test manual summary generation
      if (adminQuery.docs.isNotEmpty) {
        print('\n🧪 Testing manual summary generation...');
        try {
          await summaryService.generateAndSendDailySummary(
            organizationId: orgId,
            targetDate: now.subtract(const Duration(days: 1)),
          );
          print('   ✅ Manual summary generation completed successfully');
        } catch (e) {
          print('   ❌ Manual summary generation failed: $e');
        }
      }

      print('   ---');
    }

    // 2. Check background service status
    print('\n⏰ Step 2: Checking background service...');
    try {
      final backgroundService = DailyBackgroundService.instance;
      print('Background service instance created successfully');

      // Test trigger functionality
      print('Testing background service trigger for first org...');
      if (orgsQuery.docs.isNotEmpty) {
        final testOrgId = orgsQuery.docs.first.id;
        await backgroundService.triggerDailySummaryForTesting(organizationId: testOrgId);
        print('Background service test trigger completed');
      }
    } catch (e) {
      print('Error with background service: $e');
    }

    // 3. Check notification system
    print('\n📬 Step 3: Checking recent notifications...');

    for (final orgDoc in orgsQuery.docs.take(2)) {
      final orgId = orgDoc.id;
      final recentNotifs =
          await firestore
              .collection('organizations')
              .doc(orgId)
              .collection('notifications')
              .where('type', isEqualTo: 'general')
              .orderBy('createdAt', descending: true)
              .limit(5)
              .get();

      print('Recent notifications for org $orgId: ${recentNotifs.docs.length}');
      for (final notif in recentNotifs.docs) {
        final notifData = notif.data();
        final title = notifData['title'] ?? 'No title';
        final createdAt = (notifData['createdAt'] as Timestamp?)?.toDate();
        final isDailySum = title.contains('Daily Notes Summary') || title.contains('Daily Summary');
        print('  - ${isDailySum ? '📋 ' : '📝 '}$title (${createdAt ?? 'No date'})');
      }
    }

    // 4. Test timezone and time calculations
    print('\n🕐 Step 4: Time and timezone analysis...');
    final now = DateTime.now();
    print('Current UTC time: $now');
    print('Current hour: ${now.hour}');
    print('Should send summary (after 8 PM): ${now.hour >= 20}');
    print('Should send fallback (after 10 PM): ${now.hour >= 22}');

    print('\n✅ Debug completed successfully!');
  } catch (e, stackTrace) {
    print('\n❌ Debug failed with error: $e');
    print('Stack trace: $stackTrace');
  }
}
